### ISSUE TITLE
Move to IPython 3.x by default

### ISSUE THREAD

**Issue by rob-c** [ganga-devs/ganga#20]
I would like to see us move to IPython 3.x by default by 6.1.13.

I think this is as working as well as the presently default IPython 0.6 implementation we're using.

It would be nice to drop the old IPython 0.6 related code once we've settled on IPython 3 as it's unlikely we will have anyone dropping back to an older version any time soon.

I would push for it for 6.1.12 but there have been a lot of changes in Core which I'd like to see tested and released first and whilst IPython shouldn't interfere with this I'd rather not risk too many things changing under the hood in 6.1.12.

Does anyone have a strong objection to this move?

Thanks,

Rob

**Comment by egede** [comment]
Hi,

I read this as Python 3 first (and not IPython) and as a result nearly 
dropped my coffee. Anyway this sounds good to me.

Ulrik

**Comment by milliams** [comment]
Hopefully your coffee is safe Ulrik!

I would if possible like to move to IPython 4 as soon as possible too. It's code base is very similar to IPython 3 but it's after the notebooks and a lot of legacy code had been removed. This means that it is more likely to be supported for longer and future upgrading will be easier. The current downside is that it required Python 2.7 and so it will have to wait on that being finalised (see issue #21).

I've been running with IPython 3 for a wile now and have not spotted any problems (I've even fixed a few old regressions).

In summary: I support the move to IPython 3 ASAP. I will look into upgrading again to IPython 4 once #21 is resolved.

**Comment by rob-c** [comment]
My only caveat is does IPython4 require users to write python 2.7 compatible scripts?

I'm in favour of moving if it allows user scripts to be Python 2.6 compatible but if we're going to require users to change their scripts to be Python 2.7+ I'd like to make this move as part of 6.2 to make the distinction of what is required clearer.

**Comment by alexanderrichards** [comment]
> In summary: I support the move to IPython 3 ASAP

ditto

**Comment by egede** [comment]
Hi Matt,

Just a comment to remind us that keeping the wrapper scripts Python 2.6 
compliant is not a show stopper for this. Those wrapper scripts will 
(should) never be executed inside an iPython environment.

Ulrik.

**Comment by alexanderrichards** [comment]
I think this is agreed then so a question it is no longer :smile:

**Comment by JacekHoleczek** [comment]
Well, I'm running Ganga on a (remote cluster "User Interface") SL 6.x machine where:
[...]$ python --version
Python 2.6.6
[..]$ ipython --version
0.13.2
What you say is that I'm out of business with the next-but-one Ganga release?

**Comment by rob-c** [comment]
Hi Jacek,

My Short answer is no.
Ganga will continue to behave the same it does before 6.1.13 and users should not be exposed to this change other than seeing the interactive prompt change to use a newer IPython version.

Long answer:
Ganga ships with a external packages such as IPython and modifies it's existing runtime to ensure it gets the correct version of these packages when they're needed.
Additionally the move to require Python 2.7 is in progress and likely will not be part of 6.1.13 due to different reasons.

Regards,
Rob

**Comment by milliams** [comment]
Strangely, it seems that IPython 3 (as well as 2 and 4) requires Python 2.7 as seen at https://github.com/ipython/ipython/blob/rel-3.2.2/setup.py#L30. I'm not sure exactly how we missed this since I remember checking 3.2.1 and having it work and then trying 4.0.0 and having it tell me that my Python version (2.6) was too old which made me assume that 4.0.0 was the first to require 2.7.

**Comment by rob-c** [comment]
Interesting, I'll do some checking about what version of Python I'm working with in the LHCb environments in that case. If this is going to cause problems with Core/Atlas I'd prefer to have this issue open until we have 2.7 support in that case, just to keep track.

**Comment by JacekHoleczek** [comment]
I'm somehow confused now. It was clear to me that IPython 3 requires python 2.7.
Also the https://github.com/ganga-devs/ganga/issues/21 thread explicitly mentions it.
The http://ipython.org/ explicitly says on the main page that the new "IPython supports Python 2.7 and 3.3 or newer. Our older 1.x series supports Python 2.6 and 3.2.".
Also, the main page of http://jupyter.readthedocs.org/en/latest/install.html explicitly says "Jupyter requires Python 2.7 or ≥ 3.3." and a moment later "If you need to use Python 2.6 or 3.2, you can find IPython 1.2 here."

**Comment by rob-c** [comment]
I'm doing some investigating into what is involved in adding in IPython 1.2.1 support in parallel to what we currently have, the API looks similar to that in 3.x+ which is much nicer than what was in 0.6 imo

**Comment by rob-c** [comment]
OK, it's a one line fix in this case.
@milliams Are you planning to move to exclusively support Python 2.7 in a backwards incompatible way? I don't imagine that we'll be using much from Python 2.7 that isn't already in Python 2.6 so adding dual 2.6/2.7 support here is relatively trivial

**Comment by milliams** [comment]
@rob-c As I mention in #38 the use of Python 2.7 will be optional for now. That means that Ganga will try to use Python 2.7 if it is available but will continue to use 2.6 (with a warning and information on installing) if that is all there is. Of course this means that it doesn't really help up since we can't use any Python 2.7 features which we would like to move to doing so soon. In a few versions time we can then move to *requiring* Python 2.7. A this point we can start using IPython 4 too.

I'm glad that there's not much of a change needed to support 1.2.1, at least that's already 9 years newer than what we had before. The IPython release schedule really sped up once 1.0 was released and they seem to be getting a major version almost every year.

**Comment by JacekHoleczek** [comment]
I'm not sure if anybody realizes it ...
Some days ago, I installed IPython 3.2.2 on my Ubuntu 14.04.3 LTS / x86_64 machine (using "pip") and "pip" additionally upgraded / installed about 20 python2.7 packages (into "/usr/local/lib/python2.7/dist-packages/", so now for some packages I have two different versions installed, one coming from the "official system distribution" and a newer one installed by the "sudo pip install 'ipython[all]<4'" call).
Assuming that IPython 3 will once require python2.7, will Ganga install missing / outdated python2.7 packages somewhere in the Ganga's subdirectories?
Note also that "pip" may not be available in the system at all. I needed to add it myself when I wanted to install the new IPython (and I actually checked that it's not installed on the remote cluster which I will need to use).
In the end, it seems to me that Ganga would need to be distributed with its own IPhython AND all relevant python modules (as one cannot rely on the operating system to provide them).

**Comment by rob-c** [comment]
At the moment Ganga DOES ship with an external IPython install. This will very likely change when we move to try and get pip support but that is down the line and I'm sure backup commands in instructions for this not being available will be looked into when we stop shipping with externals.

As it stands there is the open question of do we ship the next version of Ganga with IPython 1.2.x or 3.2.x?

**Comment by alexanderrichards** [comment]
we want to try and move *away* from shipping our own versions of things with python. Ideally Ganga could be installed with pip and all dependencies worked out this way.

**Comment by milliams** [comment]
The end aim is to have the install process create a virtualenv and `pip install ganga` inside that. That will automatically install all the dependencies into the virtualenv without any clash with the system libraries.

Further down the road we intend on only depending on public libraries which are available from PyPI and are installable with `pip`.

`pip` is assumed to be standard tool and is effectively part of the standard distribution as of [Python 2.7.9](https://docs.python.org/2/library/ensurepip.html) and 3.4. It is always possible to install `pip` locally, without any need for root access.

**Comment by milliams** [comment]
@rob-c I suggest that for now we ship with 1.2.1 until we have a hard dependency on Python 2.7.

**Comment by rob-c** [comment]
ok, that's fine with me, it's still a huge step forward imo

**Comment by JacekHoleczek** [comment]
I've got an additional idea ... when the install process creates an appropriate "virtualenv" for python2.7, could it also check the existence of appropriate "pip" / "pip2" (or maybe "pip2.7" as "pip" / "pip2" may be for the default python 2.6.6) and if not found then install it locally inside of this "virtualenv" so that later it could be used to install IPython?
Note: "pip" became a default in python 2.7.9, but the majority of "production" systems use older versions (e.g. I have one with 2.7.6 and another one with 2.7.6rc1, both without "pip").

**Comment by rob-c** [comment]
Tbh we're getting off topic here so I'm looking to move the discussion back to updating IPython in Ganga 6.1.14.

We're going to introduce IPython 1.2.1 support in Ganga 6.1.14 and we have support for future IPython versions (I'll have to test if there are any peculiarities for IPython 2.x) which we get 'for free' in moving to the new API.

I'll leave this tag open for now as I would like to see if it's possible to keep the IPython history from Ganga 6.1.13 as a lot of LHCb users make extensive use of this. If this has to be lost as part of the update then we have to pay the price for stagnating on 0.6.13 for so long.

**Comment by rob-c** [comment]
Hi all,

Whilst this work has been mostly finished I'm keen to implement this properly as possible,

As a small addendum it appears that modern IPython makes use of a binary history in sqlite to store the history from the interactive prompt.
With this in mind I'm going to write a helper function into the bootstrap that will roll-forward the old text-based history into the IPython history.
(using this http://stackoverflow.com/questions/17651905/recovering-importing-missing-ipython-history-after-an-upgrade as a reference)

Given we're dropping support for IPython pre 1.2.1 in this release I think it's sensible to have this run the first time that the user launches 6.1.14 alongside the code which displays the release notes. (I'll add some logic to make sure that this is only run once).

Hopefully I'll be able to get this in for the 6.1.14 release.

Thanks,

Rob
